src/number.rs

```diff
@@ -31,6 +31,18 @@ enum N {
     Float(f64),
 }
 
+/// Describes the kind of number we're dealing with
+/// This is a hint about how the number is stored internally
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum NumberKind {
+    /// The number is a positive integer, and can be stored in an u64
```

Comment:
I would prefer not to make the guarantee that all positive integers can be stored in u64. If we move to an arbitrary precision representation, something can be a positive integer without fitting in u64. Same for the negative integer and float cases.